docs/features/nullable-metadata.md

```diff
@@ -0,0 +1,130 @@
+        AttributeTargets.Event |
+        AttributeTargets.Field |
+        AttributeTargets.GenericParameter |
+        AttributeTargets.Parameter |
+        AttributeTargets.Property |
+        AttributeTargets.ReturnValue,
+        AllowMultiple = false)]
+    public sealed class NullableAttribute : Attribute
+    {
+        public readonly byte[] NullableFlags;
+        public NullableAttribute(byte b)
+        {
+            NullableFlags = new byte[] { b };
+        }
+        public NullableAttribute(byte[] b)
+        {
+            NullableFlags = b;
+        }
+    }
+}
+```
+
+Each type reference is accompanied by a `NullableAttribute` with an array of bytes, where 0 is oblivious, 1 is not annotated and 2 is annotated.
+
+Value types may be marked with 0 (oblivious) or 1 (not annotated) - the representations are considered equivalent.
+
+To optimize trivial cases the attribute may be omitted or replaced with an attribute that takes a single byte value rather than an array. The C#8 compiler uses the following optimizations.
+
+Optimized cases:
+1. All parts are oblivious in oblivious [emit context](#NullableContextAttribute): the `NullableAttribute` is omitted
+2. All parts are oblivious in not annotated [emit context](#NullableContextAttribute): `NullableAttribute(0)` (rather than `new byte[] { 0 }`)
+3. All parts are not annotated in not annotated [emit context](#NullableContextAttribute): the `NullableAttribute` is omitted
+4. All parts are not annotated in oblivious [emit context](#NullableContextAttribute): `NullableAttribute(1)` (rather than `new byte[] { 1 }`)
+5. All parts are annotated: `NullableAttribute(2)` (rather than `new byte[] { 2 }`)
+
+The first case, where all parts are oblivious in an oblivious context, matches the representation in legacy assemblies where there are no `NullaleAttribute` annotations.
+
+The following `NullableAttribute` cases are recognized and emitted for type parameter definitions:
+1. `notnull` constraint in oblivious [emit context](#NullableContextAttribute): `NullableAttribute(1)`
+2. `class` constraint in `#nullable enable` and in oblivious [emit context](#NullableContextAttribute): `NullableAttribute(1)`
+2. `class` constraint in `#nullable disable` and in not annotated [emit context](#NullableContextAttribute): `NullableAttribute(0)`
+3. `class?` constraint: `NullableAttribute(2)`
+4. No `notnull`, `class`, `struct`, `unmanaged`, or type constraints in `#nullable enable` (the equivalent of an `object?` constraint): `NullableAttribute(2)`
+
+Other forms of `NullableAttribute` on type parameter definitions will be ignored silently.
+
+The type declaration for `NullableAttribute` is synthesized by the compiler if not already included in the compilation.
+
+Metadata examples:
+```C#
+[Nullable(2)] string s;      // string?
+[Nullable({ 2, 1, 2 }] Dictionary<string, object> d; // Dictionary<string!, object?>?
+[Nullable({ 2, 1 }] int[] a; // int[]?
+[Nullable(1)] int[] b;       // int[]!
```

Comment:
I think today we emit [Nullable({ 1, 0 }]. Are you proposing a change?

Reply:
In a `NullableContext(true)` emit context, we should use 1 for value types.

---
In reply to: [290513550](https://github.com/dotnet/roslyn/pull/36152#discussion_r290513550) [](ancestors = 290513550)

Reply:
> In a NullableContext(true) emit context, we should use 1 for value types.

Why do we need to emit any attribute for this type in ```NullableContext(true) ```?

---
In reply to: [290519517](https://github.com/dotnet/roslyn/pull/36152#discussion_r290519517) [](ancestors = 290519517,290513550)